src/lib-imap/imap-capability-list.h

```diff
@@ -0,0 +1,83 @@
+#ifndef IMAP_CAPABILITY_LIST_H
+#define IMAP_CAPABILITY_LIST_H
+
+#include <stdint.h>
+#include <stddef.h>
+
+#include "lib.h"
+#include "array.h"
+
+/* bitflags to represent different conditions of
+   the current imap session */
+typedef enum {
+	CAP_NEVER       = 0,         /* never show this capability */
+	CAP_PREAUTH     = (1 << 0),  /* show this cap before login */
+	CAP_POSTAUTH    = (1 << 1),  /* show this cap after login */
+	CAP_TLSACTIVE   = (1 << 2),  /* show this cap if tls is active */
+	CAP_TLSINACTIVE = (1 << 3),  /* show this cap if tls is inactive */
+	CAP_SSLACTIVE   = (1 << 4),  /* show this cap if ssl is active */
+	CAP_SSLINACTIVE = (1 << 5),  /* show this cap if ssl is inactive */
+	CAP_NO_LOGIN    = (1 << 6),  /* show this cap if login is disabled */
+	/* NOTE: connections from localhost are 'secure' regardless of SSL/TLS! */
+	CAP_SECURE      = (1 << 7),  /* show this cap if we are secured */
+	CAP_INSECURE    = (1 << 8),  /* show this cap if we aren't secured */
+	CAP_ALWAYS      = 0x7FFFFFFF /* always show this cap */
+} capability_cond_t;
```

Comment:
Dovecot typically doesn't use typedefs except for a few of its core library things. Also using CAP_ prefix isn't descriptive enough and could be confused with CAP_* defines in linux/capability.h. I think could be:

enum imap_capability_visiblity {
  IMAP_CAPABILITY_VISIBILITY_...

More verbose, but clearer. Perhaps IMAP_CAP_VISIBILITY_* or maybe there's a better name than "visibility". But still, "imap_" prefix at least and more verbosity would be better.

Reply:
Interesting, one of the comments by jeffpc actually asked to switch this to a typedef for the sake of gdb pretty-printing the values of variables. 

I have adjusted it to the enum you described (enum imap_capability_visibility) and I think your suggestion of IMAP_CAP_VISIBILITY_ is much better than the current CAP_ prefix, I have also updated that to the more verbose prefix.